idx/metric_definition.go

```diff
@@ -0,0 +1,552 @@
+		invalidTag.Inc()
+		return TagKeyValue{}
+	}
+	eqPos := strings.Index(tag, "=")
+	if eqPos < 0 {
+		log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+		invalidTag.Inc()
+		return TagKeyValue{}
+	}
+	key, err := IdxIntern.AddOrGet([]byte(tag[:eqPos]), true)
+	if err != nil {
+		log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+		internError.Inc()
+		return TagKeyValue{}
+	}
+	tkv.Key = key
+
+	value, err := IdxIntern.AddOrGet([]byte(tag[eqPos+1:]), true)
+	if err != nil {
+		log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+		internError.Inc()
+		return TagKeyValue{}
+	}
+	tkv.Value = value
+	return tkv
+}
+
+type KeyValuesSlice []TagKeyValue
+
+func (kvs KeyValuesSlice) Len() int      { return len(kvs) }
+func (kvs KeyValuesSlice) Swap(i, j int) { kvs[i], kvs[j] = kvs[j], kvs[i] }
+func (kvs KeyValuesSlice) Less(i, j int) bool {
+	k1, _ := IdxIntern.GetStringFromPtr(kvs[i].Key)
+	k2, _ := IdxIntern.GetStringFromPtr(kvs[j].Key)
+	return k1 < k2
+}
+
+// TagKeyValues stores a slice of all of the Tag Key/Value pair combinations for a MetricDefinition
+type TagKeyValues struct {
+	KeyValues KeyValuesSlice
+}
+
+// Strings returns a slice containing all of the Tag Key/Value pair combinations for a MetricDefinition.
+// Each item in the slice is in the form of 'key=value'
+func (t *TagKeyValues) Strings() []string {
+	tags := make([]string, len((*t).KeyValues))
+	for i, tag := range (*t).KeyValues {
+		tags[i] = tag.String()
+	}
+	return tags
+}
+
+func (t *TagKeyValues) ExtensionType() int8 {
+	return 90
+}
+
+func (t *TagKeyValues) Len() int {
+	var total int
+	for _, kv := range t.Strings() {
+		if strings.HasPrefix(kv, "name") {
+			continue
+		}
+		total += len(kv) + 1
+	}
+	return total
+}
+
+func (t *TagKeyValues) MarshalBinaryTo(b []byte) error {
+	var total int
+	for _, kv := range t.Strings() {
+		if strings.HasPrefix(kv, "name") {
+			continue
+		}
+		if kv == "=" || kv == "" {
+			continue
+		}
+		copy(b[total:], kv)
+		total += len(kv)
+		copy(b[total:], ";")
+		total++
+	}
+	return nil
+}
+
+func (t *TagKeyValues) UnmarshalBinary(b []byte) error {
+	if len(b) < 3 {
+		return nil
+	}
+	tags := strings.Split(string(b[:len(b)-1]), ";")
+	tmp := make([]TagKeyValue, len(tags))
+	t.KeyValues = tmp
+	for i, tag := range tags {
+		if strings.HasPrefix(tag, "name") {
+			continue
+		}
+		if tag == "=" || tag == "" {
+			log.Error("idx: Empty tag, ignoring: ", tag)
+			invalidTag.Inc()
+			continue
+		}
+		if strings.Contains(tag, ";") {
+			log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+			invalidTag.Inc()
+			continue
+		}
+		eqPos := strings.Index(tag, "=")
+		if eqPos < 0 {
+			log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+			invalidTag.Inc()
+			continue
+		}
+		key, err := IdxIntern.AddOrGet([]byte(tag[:eqPos]), true)
+		if err != nil {
+			log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+			internError.Inc()
+			continue
+		} else {
```

Comment:
this else block is not necessary because the end of the `if` block is a `continue`

Reply:
Yeah, no idea why that is there